### ISSUE TITLE
Capture the decode step as a bucket ladder of replayable CUDA graphs

### ISSUE THREAD

**Issue by jorgeantonio21** [atoma-network/atoma-infer#186]
## Problem Statement

Every decode step is enqueued kernel by kernel from the host. Measured on one H100 PCIe at compute capability 9.0 with a four-layer, 8B-shaped step: host enqueue costs 227–418 µs median and up to 755 µs p99 per step, against roughly 2.1–2.6 ms of device time. Host enqueue scales with layer count while device time scales with work, so at production depth the host becomes the limit and the device idles between decode steps.

Replaying a captured CUDA graph enqueues the same step in 3.6–9.9 µs median and 6.5–16.8 µs p99 — 25 to 40 times less host time.

Capture and replay of a single decode step is proven on SM90. Eight configurations — buckets 1, 8, 32 and 64, each with and without an in-graph single-rank all-reduce — all captured cleanly at 79–91 nodes in under 1 ms, produced logits bit-identical to eager execution across 32 steps of varying token ids, sequence lengths and block tables, and survived 1000 consecutive replays with zero free-memory delta.

What does not exist is the serving machinery around that single capture: choosing which captured graph serves a live batch, padding a batch up to a bucket, laying out activations so their addresses are stable by construction, declaring where a forward pass must leave the graph, and the gates that establish the result is correct and fast enough to ship.

## Solution

A decode executor that captures one CUDA graph per bucket during startup and replays the matching graph for every live batch, over an engine-owned arena whose addresses are a pure function of bucket, layer and role.

Six parts:

1. **Arena.** Activation slot offsets assigned by greedy earliest-fit over declared role lifetimes, with a no-reuse layout retained as the reference oracle.
2. **Dispatcher.** Graph key, bucket ladder, dense padding lookup table, and an admission predicate returning a named reason for every rejection.
3. **Attention contract.** A host-side preparation call, a record-only call, and a declared support level per backend.
4. **Three-phase executor.** Allocation, Capture, Replay — each phase permitting only the operations it is named for.
5. **Segmented capture.** Bridge buffers as arena roles, with break points declared by backends and models.
6. **Gates.** Correctness, memory, and throughput, measured against the existing benchmark harness.

## Requirements

**Provenance**

1. Every numeric default must carry its provenance: measured on named hardware, or computed from a stated formula.
2. Numbers measured at one model shape must not be presented as valid at another. Where a constant is known to scale with model depth or width, the scaling relationship must be stated alongside it.

**Arena and activation layout**

3. Activation addresses must be a pure function of bucket, layer and role, exposed only as a lookup and never as a formula at a call site.
4. One arena must serve every bucket. Buckets never replay concurrently, so the arena is sized at the largest bucket's footprint, not the sum across the bucket ladder.
5. Each role must be declared with a per-token width in bytes and a lifetime: the half-open range of the layer's op order in which its slot holds live data.
6. Slot offsets must be assigned by greedy earliest-fit over declared lifetimes. Two roles whose lifetimes do not overlap may share bytes; two whose lifetimes overlap must not.
7. The arena must reject a role table containing an empty or inverted lifetime, naming the offending role.
8. A no-reuse layout — one slot per layer, no sharing — must remain selectable as the reference the reuse layout is checked against.
9. A poison mode must fill each slot with a signalling pattern before its first use and after its last use, so a read outside a role's lifetime produces deterministic garbage rather than plausible numbers.
10. The bucket ladder must enter the arena uninterpreted: never sorted, deduplicated, or assumed monotonic. A bucket's addressing must not depend on its position in the list.
11. The arena must hold only what is sized per bucket: activations and bridge buffers.
12. Buffers sized once at ladder maximum — per-step inputs, outputs, and kernel workspaces — must be owned by the graph entry whose graph baked their addresses, not by the arena.
13. Total arena size and each bucket's footprint must be computable in closed form from layer count, role widths and the bucket ladder, with no device present.
14. The arena must carry no model knowledge. Roles enter as caller-declared widths and lifetimes, indexed positionally.

**Graph key and dispatch**

15. The graph key must be built by exactly one pure function of the current padded batch.
16. The key must consist of padded token count, request count, and a uniform-decode flag.
17. The key type must have no default value, no public fields, and exactly one constructor, so that adding a field later is a compile error at the construction site rather than a silent default.
18. Keys must bind the exact request count, not only the padded token count.
19. A dense lookup table must map any token count to the next captured bucket.
20. The dispatcher must own dispatch truth. Executors must act on its result without re-deriving it.
21. Dispatch priority must be full-graph replay, then segmented replay, then eager.
22. The bucket ladder must be fully config-exposed as a list of bucket sizes.
23. The default ladder must be `[1, 2, 4]`, then 8 to 128 in steps of 8, then 128 to the maximum in steps of 64.
24. The ladder maximum must be per-platform, defaulting to 512, and to 1024 on data-center Blackwell.

**Padding**

25. A batch smaller than its bucket must be padded up, not dropped to eager.
26. Padding must use persistent dummy requests owning real KV blocks preallocated during Allocation. Allocating dummy KV lazily races cache saturation into permanent eager fallback.
27. There must be enough dummy slots to fill any bucket from any admitted request count, each owning its own KV blocks. A single shared dummy cannot fill a bucket that binds an exact request count.
28. Sentinel sequence lengths must not be used. The real-KV dummy must be the only padding mechanism.
29. A device-resident count of non-padded tokens must be available to kernels that need it.
30. The total dummy-KV cost implied by a configuration must be computable and reported at configuration time.

**Admission**

31. Admission must be a predicate returning either a graph key or a named rejection reason carrying the concrete numbers that caused it.
32. Rejection reasons must be a closed set: padded token count above ladder maximum; request count above the captured maximum; backend support level insufficient for the batch shape; batch is not uniform decode.
33. Eager fallbacks must be counted and logged per reason.
34. There must be no automatic recapture at runtime, for any reason.

**Attention contract**

35. Attention backends must expose a host-side preparation call that may allocate and synchronize, and that re-plans scheduling metadata at fixed device addresses before every replay.
36. Attention backends must expose a record call issuing static-shape device work only: no host readback, no synchronization, no dynamic allocation.
37. Backends must declare a support level from: always; uniform batch; uniform single-token decode; never.
38. Support levels must describe the validity of the captured routine, not kernel capability.
39. The graph mode must be downgraded at startup to the minimum support level across all active backends.
40. Kernels reachable from a captured step must accept a caller-owned workspace and must not allocate.
41. Eager and captured workspaces must be structurally separate.

**Break points**

42. A backend must be able to declare a break point as a capability statement: an op it cannot capture, or an op that is rank-coupled.
43. A model definition must be able to declare a break point as a policy statement, marking where its eager region lies.
44. The dispatcher must take the union of backend-declared and model-declared break points.
45. Break points must not be hardwired to the attention op.

**Executor phases**

46. The executor must have three phases — Allocation, Capture, Replay — each permitting only the operations it is named for.
47. Every device allocation, stream, BLAS handle and communicator must be created during Allocation.
48. Obtaining an owning reference to the underlying stream must be possible only during Allocation.
49. Beginning a capture must be impossible during Replay by construction, not by a runtime flag.
50. An explicit weight-reload path must be able to return from Replay to Allocation for re-warmup. No other path may.
51. One eager warmup pass must run before any capture, to settle lazy state in libraries that allocate or select algorithms on first use.
52. One eager warmup run at the graph's exact shape must run immediately before each capture.
53. Capture must record on a dedicated side stream in relaxed mode, never the default stream.
54. Graph instantiation must use flags zero.
55. Enqueue onto the capture stream must go through a single descriptor seam. Call sites must not handle raw stream handles.
56. Device-event tracking must be disabled globally at engine start, not per capture.
57. An invalidated capture must be discardable without instantiating.
58. Free device memory must be logged per captured graph, and total capture time reported as a startup cost.

**Per-step replay**

59. Per-step copy-in must be a single fused pass.
60. Block tables must be staged at full width into a fixed-address tensor, never at a host-computed column cap.
61. Sequence and KV lengths must be host-native in the scheduler, so no planning call hides a device-to-host synchronization.
62. Sampled tokens must be device-resident, with the next step's input ids gathered on device.
63. Pinned staging must be a ring of configurable depth, defaulting to two.
64. An event fence must guard reuse of a pinned staging entry. The host must not overwrite an entry until the copy reading it has completed.
65. Dummy runs must participate in the fence protocol.
66. Outputs must live in graph-owned memory, handed out as non-owning views sliced to the real batch.
67. Debug builds must assert device-pointer equality for every baked address at every replay.

**Collectives**

68. All communicators must be created before any capture.
69. Collectives issued during capture must be recorded into the graph.
70. Teardown order must be buffers, then graph executable, then graph, then communicator.
71. Elastic membership changes must trigger full re-warmup, never graph patching.

**Segmented capture**

72. A forward pass must be capturable as an ordered list of segments and eager regions, and replayed in that order.
73. Bridge buffers must be arena roles whose lifetime spans the break point.
74. An eager operation between segments must write in place into the bridge buffer and must not return a fresh allocation.
75. No allocation may occur in an eager region between segments.
76. A rank-coupled break point must skip its body during capture, recording only the output address, and must re-synchronize ranks before the eager region.
77. Side streams forked inside a segment must be joined before the segment ends.
78. Side-effecting hooks must run in eager regions.

**Gates**

79. Every bucket in the ladder must capture successfully.
80. Replayed logits must be bit-identical to eager for at least 32 steps of varying token ids, sequence lengths and block tables.
81. Every baked address must be unchanged after each replay.
82. 1000 consecutive replays must produce zero free-memory delta.
83. The reuse layout must produce logits bit-identical to the no-reuse layout.
84. Graph memory must be reported against the KV budget as a fixed term plus a marginal term per graph.
85. Profiling must show no device idle between decode steps.
86. Host time per step must stay below 200 µs p99. This is met by more than an order of magnitude under replay and unreachable eagerly at bucket 64; it functions as a regression check, not as a tuning target.
87. Throughput must reach at least 75% of the pinned vLLM baseline on an 8B model at matched hardware and configuration.
88. The same throughput gate must be defined for the frontier-MoE target and recorded as not yet measured.
89. Goodput at a fixed SLO must be measured across arena layouts, establishing that freed bytes become tokens.
90. Gates must run on the GPU verification path, with nightly regression against the pinned baseline.

## Implementation Decisions

**Crate boundaries.** `atoma-runtime` owns the capture lifecycle, the arena, graph entries and the executor API; it links the driver and carries no model knowledge. A new crate, `atoma-core`, owns the graph key, the dense lookup table, the bucket ladder, the admission predicate and the dispatcher; it links no driver and is fully testable without a GPU. The dispatcher work creates `atoma-core`.

The boundary is drawn on what can be verified without a device, not on subject matter. The dispatcher is pure arithmetic and belongs on the GPU-free side.

**Arena interface.** Unchanged in shape: a lookup from bucket, layer and role to a byte offset. Construction takes the layer count, the role table and the bucket ladder. Only the offset-table computation changes — from one slot per layer to greedy earliest-fit over lifetimes. No caller changes.

Role declaration carries the lifetime:

```
RoleSpec { width_bytes, first_use, last_use }
```

where `first_use` and `last_use` index a caller-declared linear op order within a layer. This is a table, not an operation graph: the caller declares the order, the arena does not derive it.

**Layout selection.** Three layouts behind one interface: greedy earliest-fit (default), no-reuse (reference oracle), and poison (debug). The reuse layout's correctness is established by bit-identity against the no-reuse layout, and localised by poison mode when it fails.

**Graph key.** A struct with private fields and one constructor taking the padded batch. No variant bits exist at this scope.

**Rejection reasons.** A closed enum, each variant carrying the numbers that caused it, so the eager-fallback counter is per-reason rather than a single opaque total.

**Executor phases.** Three types, with consuming transitions. The owning stream reference is reachable only from the Allocation type, so binding a BLAS handle or a communicator after capture has begun is not expressible. Replay has one consuming transition back to Allocation, used only by weight reload.

**Enqueue seam.** A descriptor trait implemented per backend. Kernels reached through C ABI take a raw stream in their signature and always will; the trait confines that to one implementation per backend rather than one occurrence per call site. This is the only path onto the capture stream.

**Staging fence.** A fence type owning the event, exposing a record call taking the capture stream and both blocking and non-blocking waits. The capture stream itself gains no synchronize.

**Graph memory model.** Affine, not multiplicative: a fixed term plus a marginal term per graph. Measured at four layers and 79–91 nodes per graph: 34 MiB fixed, approximately 237 KiB marginal per graph, amortised across 96 distinct buckets, with driver allocations landing at 2 MiB granularity. The marginal term is expected to scale with node count and must be re-measured at production depth before it is used in a budget.

**Arena economics.** At production dimensions and a bucket maximum of 1024, one slot per layer reaches the multi-gigabyte range while lifetime-based reuse reaches the low hundreds of megabytes. The exact figures are computed as part of this work, since they are closed-form and require no device.

**Capture ordering.** No ordering constraint. Capturing largest-bucket-first is a memory-pool optimisation; instantiation with flags zero over a preallocated arena has no pool, and the arena is sized at the largest bucket by construction regardless of capture order. Measured across 96 buckets captured in ascending order, the memory curve is affine with no ordering effect.

## Testing Decisions

A good test here asserts external behaviour: an address, a dispatch outcome, a rejection reason, a measured number. A test that asserts how an offset table was computed, or which internal helper ran, is testing implementation and will break under the layout changes this work explicitly plans for.

**Seam 1 — the arena.** Extends the existing arena unit tests, which already cover hand-computed layouts, slot non-overlap, footprint totals, ladder-order independence and out-of-range rejection. New coverage: lifetime-based offsets against hand-computed layouts; the property that no two roles with overlapping lifetimes share bytes, over generated role tables; rejection of empty and inverted lifetimes; bridge-buffer lifetimes spanning a break point; the no-reuse and poison layouts; and the closed-form size comparison across layouts at production dimensions. All GPU-free.

**Seam 2 — the dispatcher.** New, in `atoma-core`. Driven by a fake backend declaring a support level and break points. Coverage: key construction from padded batches; the padding lookup table across the full token range including boundaries at every bucket edge; each of the four rejection reasons triggered with its numbers; per-reason fallback counting; the minimum-support-level downgrade across several backends; and the union of backend-declared and model-declared break points. All GPU-free.

**Seam 3 — the benchmark harness.** Existing. Carries the empirical gates using its goodput-at-SLO measurement, its KV-capacity probe, and its pinned vLLM baseline comparison: the arena-layout goodput A/B, and the throughput ratio gate.

**GPU evidence.** Bit-identity, address stability, the replay soak, per-graph memory, and every segmented-capture criterion run as feature-gated tests on the GPU verification path, following the existing pattern for feature-gated engine tests.

**Failure verification.** Each new gate must be shown to fail before it is trusted: break the layout, confirm bit-identity fails; declare a lifetime one op short, confirm poison mode catches it.

## Out of Scope

- The runtime tensor type and swapping model definitions onto it.
- The FlashInfer SM90 attention swap and the deletion of the current attention sources.
- The engine core: pinned engine thread, token-budget scheduler, and the engine-to-executor ring.
- Speculative decoding, LoRA, FP8, and guided decoding. Their interaction with the graph key is anticipated but not built.
- Custom IPC all-reduce and NCCL symmetric-memory windows.
- Runtime recapture of any kind.
- Automatic derivation of activation lifetimes from an operation graph. Lifetimes are declared.
- Per-step graph executable update.
- Per-layer role tables. Roles are declared once for the model.

## Further Notes

Anticipated and deliberately absent until there is something to bind them to: variant bits in the graph key for quantization mode, attention short and long modes, and sampler variant; multi-rank key bits, which when they exist must be typed as group-synced values; bucket sizes rounded to multiples of the speculative verify length; and per-layer-class role tables for models whose early layers differ structurally from their later ones.

Graph executable update is reserved for repairing a segment after re-capture, and for optional executable deduplication across shapes. It is not a per-step mechanism.

Measurements that do not yet exist, and which gates above depend on: capture time per graph at production depth; marginal graph memory at production depth; collectives under capture at world size greater than one; and segmented capture of any kind.
